File: python/X2Capture2D.py
# Copyright (c) RVBUST, Inc - All rights reserved.
import PyRVC as RVC
import os
import numpy as np
import cv2


# Create a directory.
def TryCreateDir(d):
    if not os.path.exists(d):
        os.makedirs(d)
        return 1
    else:
        return 0


def App():

    # Initialize RVC X system.
    RVC.SystemInit()

    # Choose RVC X Camera type (USB, GigE or All)
    opt = RVC.SystemListDeviceTypeEnum.All

    # Scan all RVC X Camera devices.
    ret, devices = RVC.SystemListDevices(opt)
    print("RVC X Camera devices number:%d" % len(devices))

    # Find whether any RVC X Camera is connected or not.
    if len(devices) == 0:
        print("Can not find any RVC X Camera!")
        RVC.SystemShutdown()
        return 1
    print("devices size = %d" % len(devices))

    # Create a RVC X Camera
    x = RVC.X2.Create(devices[0])

    # Test RVC X Camera is valid or not.
    if x.IsValid() == True:
        print("RVC X Camera is valid!")
    else:
        print("RVC X Camera is not valid!")
        RVC.X2.Destroy(x)
        RVC.SystemShutdown()
        return 1

    # Open RVC X Camera.
    ret1 = x.Open()

    # Test RVC X Camera is opened or not.
    if ret1 and x.IsOpen() == True:
        print("RVC X Camera is opened!")
    else:
        print("RVC X Camera is not opened!")
        RVC.X2.Destroy(x)
        RVC.SystemShutdown()
        return 1

    # Create saving address of image and point map.
    save_address = "Data"
    TryCreateDir(save_address)

    # capture option
    cap_opt = RVC.X2_CaptureOptions();
    cap_opt.exposure_time_2d = 50
    # X2 Capture 2D
    ret2 = x.Capture2D(RVC.CameraID_Left, cap_opt)
    if ret2 == True:
        print("Capture 2D success!")
        img = x.GetImage(RVC.CameraID_Left)
        # Convert image to array and save it.
        img = np.array(img, copy=False)
        cv2.imwrite("Data/test_2d.png", img)
        print("Save image successed!")
    else:
        print("Capture 2D failed!")
        
    # X2 Capture 3D
    ret3 = x.Capture()
    if ret3 == True:
        print("RVC X Camera capture successed!")

        # Get image data and image size.
        img = x.GetImage(RVC.CameraID_Left)
        width = img.GetSize().cols
        height = img.GetSize().rows

        # Convert image to array and save it.
        img = np.array(img, copy=False)
        cv2.imwrite("Data/test.png", img)
        print("Save image successed!")

        # Convert point map (m) to array and save it.
        pm = np.array(x.GetPointMap(), copy=False).reshape(-1, 3)
        np.savetxt("Data/point_map.xyz", pm)
        print("Save point map successed!")
    else:
        print("RVC X Camera capture failed!")
        x.Close()
        RVC.X2.Destroy(x)
        RVC.SystemShutdown()
        return 1

    # Close RVC X Camera.
    x.Close()

    # Destroy RVC X Camera.
    RVC.X2.Destroy(x)

    # Shutdown RVC X System.
    RVC.SystemShutdown()

    return 0


if __name__ == "__main__":
    App()

File: c++/X2Capture2D.cpp
#include <RVC/RVC.h>
#include <iostream>
#include "IO/FileIO.h"

int main(int argc, char* argv[]) {
    // Initialize RVC X system
    RVC::SystemInit();

    // Scan all RVC X Camera devices
    RVC::Device devices[10];
    size_t actual_size = 0;
    SystemListDevices(devices, 10, &actual_size, RVC::SystemListDeviceType::All);

    // Find whether any RVC X Camera is connected or not
    if (actual_size == 0) {
        std::cout << "Can not find any RVC X Camera!" << std::endl;
        return -1;
    }

    // Create a RVC X Camera
    RVC::X2 x2 = RVC::X2::Create(devices[0]);

    // Open RVC X Camera
    x2.Open();

    // Test RVC X Camera is opened or not
    if (!x2.IsOpen()) {
        std::cout << "RVC X Camera is not opened!" << std::endl;
        RVC::X2::Destroy(x2);
        RVC::SystemShutdown();
        return 1;
    }

    const std::string save_directory = "./Data/";
    MakeDirectories(save_directory);

    RVC::X2::CaptureOptions cap_opt;
    cap_opt.exposure_time_2d = 50;
    // X2 Capture 2D with left camera
    if (x2.Capture2D(RVC::CameraID_Left, cap_opt)) {
        std::cout << "Capture 2D success!" << std::endl;
        RVC::Image img = x2.GetImage(RVC::CameraID_Left);
        std::string img_addr = save_directory + "test_2d.png";
        std::cout << "Save image to file: " << img_addr << std::endl;
        img.SaveImage(img_addr.c_str());
    } else {
        std::cout << "Capture 2D failed!" << std::endl;
    }

    // X2 Capture 3D
    if (x2.Capture()) {
        std::cout << "Capture 3D successed!" << std::endl;
    } else {
        std::cout << "Capture 3D failed!" << std::endl;
    }

    // Close RVC X Camera
    x2.Close();

    // Destroy RVC X Camera
    RVC::X2::Destroy(x2);

    // Shut Down RVC X System
    RVC::SystemShutdown();
    return 0;
}